pkg/sr/serde.go

```diff
@@ -316,44 +312,103 @@ func (s *Serde) DecodeNew(b []byte) (any, error) {
-		br := io.ByteReader(&r)
-		l, err := binary.ReadVarint(br)
-		if l == 0 { // length 0 is a shortcut for length 1, index 0
-			t = t.subindex[0]
-		}
-		for err == nil && t.subindex != nil && l > 0 {
-			var idx int64
-			idx, err = binary.ReadVarint(br)
-			t = t.subindex[int(idx)]
-			l--
-		}
+		var index []int
+		index, b, err = h.DecodeIndex(b)
 		if err != nil {
-			return nil, t, err
+			return nil, tserde{}, err
+		}
+		for _, idx := range index {
+			if t.subindex == nil {
+				return nil, tserde{}, ErrNotRegistered
+			}
+			t = t.subindex[idx]
 		}
-		b = r.b
 	}
 	if !t.exists {
-		return nil, t, ErrNotRegistered
+		return nil, tserde{}, ErrNotRegistered
 	}
 	return b, t, nil
 }
 
-type bReader struct{ b []byte }
+// SerdeHeader encodes and decodes a message header.
+type SerdeHeader interface {
+	AppendEncode(b []byte, id int, index []int) ([]byte, error)
+	DecodeID(in []byte) (id int, out []byte, err error)
+	DecodeIndex(in []byte) (index []int, out []byte, err error)
+}
 
-func (b *bReader) ReadByte() (byte, error) {
-	if len(b.b) > 0 {
-		r := b.b[0]
-		b.b = b.b[1:]
-		return r, nil
+// ConfluentHeader is a SerdeHeader that produces the Confluent wire format. It
+// starts with 0, then big endian uint32 of the ID, then index (only protobuf),
+// then the encoded message.
+//
+// https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format
+type ConfluentHeader struct{}
+
+// AppendEncode appends an encoded header to b according to the Confluent wire
+// format and returns it. Error is always nil.
+func (ConfluentHeader) AppendEncode(b []byte, id int, index []int) ([]byte, error) {
+	b = append(
+		b,
+		0,
+		byte(id>>24),
+		byte(id>>16),
+		byte(id>>8),
+		byte(id>>0),
+	)
+
+	if len(index) > 0 {
+		if len(index) == 1 && index[0] == 0 {
+			b = append(b, 0) // first-index shortcut (one type in the protobuf)
+		} else {
+			b = binary.AppendVarint(b, int64(len(index)))
+			for _, idx := range index {
+				b = binary.AppendVarint(b, int64(idx))
+			}
+		}
+	}
+
+	return b, nil
+}
+
+// DecodeID strips and decodes the schema ID from b. It returns the ID alongside
+// the unread bytes. If the header does not contain the magic byte or b contains
+// less than 5 bytes it returns ErrBadHeader.
+func (ConfluentHeader) DecodeID(b []byte) (int, []byte, error) {
+	if len(b) < 5 || b[0] != 0 {
+		return 0, nil, ErrBadHeader
+	}
+	id := binary.BigEndian.Uint32(b[1:5])
+	return int(id), b[5:], nil
+}
+
+// DecodeIndex strips and decodes the index from b. It returns the index
+// alongside the unread bytes. It expects b to be the output of DecodeID (schema
+// ID should already be stripped away).
+func (ConfluentHeader) DecodeIndex(b []byte) ([]int, []byte, error) {
+	buf := bytes.NewBuffer(b)
+	l, err := binary.ReadVarint(buf)
+	if err != nil {
+		return nil, nil, err
+	}
+	if l == 0 { // length 0 is a shortcut for length 1, index 0
+		return []int{0}, buf.Bytes(), nil
+	}
+	index := make([]int, l)
```

Comment:
This trust the value to be formatted correctly too much -- if somebody has a busted encoder and `l` here is unreasonably large, this will crash the program. I'm going to pull this branch and modify it a little bit (which also will make it easier to see the diff here). I think Header can be a SerdeOpt as well, passed to NewSerde, to set the default -- and change it per type (not that anybody would).

Reply:
FWIW it _may_ be ok to allocate blindly, but I want to play around with the API a bit here to see if something else feels good. The old code iteratively advances into the index so it doesn't allocate everything per what's on the wire.

Reply:
I think a way to solve it would be to have a maxLength field, and then to keep the max index length per registered ID.

Reply:
> I think a way to solve it would be to have a maxLength field, and then to keep the max index length per registered ID.

Have a look at the [last commit](https://github.com/twmb/franz-go/pull/467/commits/b107644c68a2a1dfcd993d416683533f92aaf93a), is that what you had in mind?